### PR TITLE
PR 2 of 3 TMB calculation revamp

### PR DESCRIPTION
**Pull request by cansavvy**
#### What scientific question is your analysis addressing?

Currently the TMB stats calculated use all mutations, but we should also have a separate TMB stats for coding mutations only, here is where I updated the denominators to be more consistent with the changes made in #307 

#### What GitHub issue does your pull request address?

Partially #305 , but also a [discussion that occurred](https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/307#discussion_r353264233) on PR #307

#### What was your approach?

Strelka and Mutect are only used for the "all mutations" TMB, but the consensus file of Strelka, Lancet, and Mutect is used for the "coding only" mutations this was discussed on #305 

### Summary of new TMB calculations
*I will add this to the README as soon as we decide we are keeping these methods. 

#### Coding only TMB:
Here exonic genome size and # exonic snvs are determined by using the methods used [here in focal CN](https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/6077149dc360843d6609cbf5e57ad5a6679ebced/analyses/focal-cn-file-preparation/01-prepare-cn-file.R#L241). SNVs outside of the exons in the gtf from Gencode 27 that is included in the download are filtered out before being summed and used for TMB calculations like such: 

This is calculated the same way for both WGS and WXS samples: 
```
coding_only_TMB = (total # exonic snvs called by all three of Strelka, Lancet, and Mutect) / exonic_genome_size 
```

#### All mutations TMB:

WXS samples are filtered using the gtf v27 file included in download and described above, 
```
WXS_all_mutations_TMB = (total # exonic snvs called by Strelka and Mutect) / exonic_genome_size 
```

For WGS samples, the size of the genome covered by the intersection of Strelka and Mutects' surveyed areas is used for the denominator:
```
WGS_all_mutations_TMB = (total # snvs called by Strelka and Mutect) / intersection_genome_size 
```

### Directions for reviewers. Tell potential reviewers what kind of feedback you are soliciting.

- I need *big picture* feedback first. 
- **Do you have ideas for quality controls for this analysis?** i.e. we *think* this makes more sense, but how do we check the validity and/or consistency of these calculations? 
- Do these denominator calculations make more scientific sense?
I've changed from using the WXS file, to using the gtf file directly for both filtering and denominator of TMB, does this make more sense? 
- I'm concerned this script is much more difficult to follow after these changes. Do we have recommendations for DRYing it up or where more documentation could help? Maybe places where things should be rearranged or reconfigured? 

#### Is the analysis in a mature enough form that the resulting figure(s) and/or table(s) are ready for review?
 The results of this analysis get sent to downstream analyses, so there are no results per se, but there would be when we carry it to `tmb-compare-tcga`. 

### Results

No new results as of now. But once we get this approved/merged, I can re-run the pipeline on AWS and obtain official TMB results files. 

#### Reproducibility Checklist

No changes to these items. 
- [x] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [x] This analysis has been added to continuous integration.

### CODE REVIEW

**Comment by jashapiro** [comment]
I haven't looked at the code yet, but I wanted to get a quick thought in before I do. 

You are using the intersection of the bed files fore WGS, which I think makes sense, and it is there that I think the intersection may be most important (though I am curious what the difference in size is between the primary bed files and the intersection... is the intersection much less than the smaller of the two?)

However, I think it would make sense to use the same intersection approach for the WXS and exon denominators:
* For the WXS it would be the intersection of the two WXS bed files (not just exons, as the WXS panel may include other sequences, and we may as well include all of that).
* For the coding changes, the numerator would be filtered to snps in the GTF ranges, but the denominator would be the intersection of the exon GTF and the WXS or WGS beds, as appropriate.

It might make sense to create these intersections using bedtools in the wrapper shell script, then feed those in as single files, rather than having the logic to deal with multiple bed files that you have to intersect inside your current script. I'm pretty sure bedtools has the ability to filter a GTF to intervals annotated as exons, but I'd have to look at the manual for that.

As I see it, there could be 4 bed files:
intersect_strelka_mutect_WGS.bed
intesect_strelka_mutect_WXS.bed
intersect_exons_strelka_mutect_lancet_WGS.bed
intersect_exons_strelka_mutect_lancet_WXS.bed

The first two would just be used for denominators, while the latter two could be used for both filtering and as the denominator (the effect on SNV counts should be identical to filtering on exons directly, which is a good sanity check).

**Comment by cansavvy** [comment]
> * For the WXS it would be the intersection of the two WXS bed files (not just exons, as the WXS panel may include other sequences, and we may as well include all of that).

There is only one WXS ranges file for all callers.

**Comment by cansavvy** [comment]
@jashapiro , when you get a chance, take another glance and see if this is more what you are talking about. I think it's easier to follow now for sure.

File: analyses/snv-callers/run_caller_consensus_analysis.sh [hunk 1]
```diff
@@ -45,15 +45,36 @@ Rscript analyses/snv-callers/scripts/02-merge_callers.R \
...
-  
+
+###################### Create intersection BED files ###########################
+# Make All mutations BED file
+bedtools intersect \
+  -a data/WGS.hg38.strelka2.unpadded.bed \
+  -b data/WGS.hg38.mutect2.unpadded.bed > scratch/intersect_strelka_mutect_WGS.bed
+
+# Convert GTF to BED file for use in bedtools
+cat data/gencode.v27.primary_assembly.annotation.gtf.gz | gunzip - | grep transcript_id | grep gene_id | convert2bed --do-not-sort --input=gtf - > scratch/gencode.v27.primary_assembly.annotation.bed
```
[the remark this answers is not in the document]
Comment (jashapiro) [reply analyses/snv-callers/run_caller_consensus_analysis.sh]: You are checking for `transcript_id` and  `gene_id`,  but not for `exon`. Do we need to filter for exons too? There are some things tagged with transcripts that are not exons... do we know whether those are in the WXS set? In a quick test, swapping `exon` for `gene_id` gave the same result as grepping for all three...

File: analyses/snv-callers/run_caller_consensus_analysis.sh [hunk 2]
```diff
@@ -45,15 +45,36 @@ Rscript analyses/snv-callers/scripts/02-merge_callers.R \
...
+
+# Make WGS coding BED file
+bedtools intersect \
+  -a data/WGS.hg38.strelka2.unpadded.bed \
+  -b data/WGS.hg38.mutect2.unpadded.bed data/WGS.hg38.lancet.300bp_padded.bed scratch/gencode.v27.primary_assembly.annotation.bed > scratch/intersect_exon_lancet_strelka_mutect_WGS.bed
```
Comment (jashapiro) [inline analyses/snv-callers/run_caller_consensus_analysis.sh]:
```suggestion
  -b data/WGS.hg38.mutect2.unpadded.bed data/WGS.hg38.lancet.300bp_padded.bed \
     scratch/gencode.v27.primary_assembly.annotation.bed \
  > scratch/intersect_exon_lancet_strelka_mutect_WGS.bed
```

File: analyses/snv-callers/run_caller_consensus_analysis.sh [hunk 3]
```diff
@@ -45,15 +45,36 @@ Rscript analyses/snv-callers/scripts/02-merge_callers.R \
 python3 analyses/snv-callers/scripts/01-setup_db.py \
   --db-file $dbfile \
   --consensus-file $consensus_file
-  
+
+###################### Create intersection BED files ###########################
+# Make All mutations BED file
+bedtools intersect \
+  -a data/WGS.hg38.strelka2.unpadded.bed \
+  -b data/WGS.hg38.mutect2.unpadded.bed > scratch/intersect_strelka_mutect_WGS.bed
```
Comment (jashapiro) [inline analyses/snv-callers/run_caller_consensus_analysis.sh]: It probably makes sense to make these scratch file names bash variables, for ease of reuse and error prevention.

File: analyses/snv-callers/run_caller_consensus_analysis.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (cansavvy) [reply analyses/snv-callers/run_caller_consensus_analysis.sh]: I'll do this test and see what happens.

[the remark this answers is not in the document]
Comment (cansavvy) [reply analyses/snv-callers/run_caller_consensus_analysis.sh]: Good call.

File: analyses/snv-callers/run_caller_consensus_analysis.sh (hunk 3 above)
Comment (cansavvy) [reply analyses/snv-callers/run_caller_consensus_analysis.sh]: Will do

File: analyses/snv-callers/run_caller_consensus_analysis.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (cansavvy) [reply analyses/snv-callers/run_caller_consensus_analysis.sh]: It's giving me the same gencode BED file whether I use `exon_id` or `gene_id`? So I think it doesn't matter? But I do wanna know if there are non-exon transcripts sneaking in here.

[the remark this answers is not in the document]
Comment (jashapiro) [reply analyses/snv-callers/run_caller_consensus_analysis.sh]: Hunh. I was pretty sure I got something shorter with grep exon. I didn’t use exon_id for whatever difference that makes.

**Comment by cansavvy** [comment]
Okay. Well this warrants some further investigation before we merge then. Will look into things.

File: analyses/snv-callers/run_caller_consensus_analysis.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (jashapiro) [reply analyses/snv-callers/run_caller_consensus_analysis.sh]:
```
rstudio@d5c7bda89fb5:~/kitematic$ gunzip -c data/gencode.v27.primary_assembly.annotation.gtf.gz   \
| grep transcript_id | grep gene_id   \
| convert2bed --do-not-sort --input=gtf -   \
> scratch/gencode.v27.primary_assembly.annotation.bed
rstudio@d5c7bda89fb5:~/kitematic$ gunzip -c data/gencode.v27.primary_assembly.annotation.gtf.gz  \
| grep transcript_id | grep exon \
| convert2bed --do-not-sort --input=gtf - \
> scratch/gencode.v27.primary_assembly.annotation.bed.exon
rstudio@d5c7bda89fb5:~/kitematic$ gunzip -c data/gencode.v27.primary_assembly.annotation.gtf.gz  \
| grep transcript_id | grep exon_id \
| convert2bed --do-not-sort --input=gtf - \
> scratch/gencode.v27.primary_assembly.annotation.bed.exon_id

rstudio@d5c7bda89fb5:~/kitematic$ wc -l scratch/gencode.v27.primary_assembly.annotation.bed
2559567 scratch/gencode.v27.primary_assembly.annotation.bed
rstudio@d5c7bda89fb5:~/kitematic$ wc -l scratch/gencode.v27.primary_assembly.annotation.bed.exon
2359079 scratch/gencode.v27.primary_assembly.annotation.bed.exon
rstudio@d5c7bda89fb5:~/kitematic$ wc -l scratch/gencode.v27.primary_assembly.annotation.bed.exon_id
2358980 scratch/gencode.v27.primary_assembly.annotation.bed.exon_id
```

So, three different filters, three different bed file results... Hooray! There are some transcripts that are not exons, some things with exon labels somewhere that don't have `exon_id`s... What joy gtf files are!

[the remark this answers is not in the document]
Comment (cansavvy) [reply analyses/snv-callers/run_caller_consensus_analysis.sh]: So is our conclusion to use `grep transcript_id | grep exon`?

[the remark this answers is not in the document]
Comment (jashapiro) [reply analyses/snv-callers/run_caller_consensus_analysis.sh]: I don’t know... do we know what the annotations that have exon but not exon_id?

File: analyses/snv-callers/run_caller_consensus_analysis.sh [hunk 4]
```diff
@@ -56,7 +56,8 @@ bedtools intersect \
   -b data/WGS.hg38.mutect2.unpadded.bed > $all_mut_wgs_bed
 
 # Convert GTF to BED file for use in bedtools
-cat data/gencode.v27.primary_assembly.annotation.gtf.gz | gunzip - | grep transcript_id | grep exon | convert2bed --do-not-sort --input=gtf - > $exon_file
+# Here we are only extracting things labeled with CCDS ids, i.e. are coded in protein
+cat data/gencode.v27.primary_assembly.annotation.gtf.gz | gunzip - | grep transcript_id | grep -P "CCDS" | convert2bed --do-not-sort --input=gtf - > $exon_file
```
Comment (jashapiro) [inline analyses/snv-callers/run_caller_consensus_analysis.sh]: I'm not 100% confident in this, but let's try using structure of the file to make this more specific:

```
gunzip -c data/gencode.v27.primary_assembly.annotation.gtf.gz | awk '$3 ~ /CDS/' | convert2bed --do-not-sort --input=gtf - > $exon_file
```
There is an interesting wrinkle that most exons occur in there twice, but I think when you merge the bed  that should be taken care of.

File: analyses/snv-callers/run_caller_consensus_analysis.sh [hunk 5]
```diff
@@ -45,15 +48,41 @@ Rscript analyses/snv-callers/scripts/02-merge_callers.R \
...
+
+###################### Create intersection BED files ###########################
+# Make All mutations BED file
+bedtools intersect \
+  -a data/WGS.hg38.strelka2.unpadded.bed \
+  -b data/WGS.hg38.mutect2.unpadded.bed > $all_mut_wgs_bed
+
+# Convert GTF to BED file for use in bedtools
+# Here we are only extracting things labeled with CCDS ids, i.e. are coded in protein
+gunzip -c data/gencode.v27.primary_assembly.annotation.gtf.gz | awk '$3 ~ /CDS/' | convert2bed --do-not-sort --input=gtf - > $exon_file
```
Comment (jashapiro) [inline analyses/snv-callers/run_caller_consensus_analysis.sh]:
```suggestion
gunzip -c data/gencode.v27.primary_assembly.annotation.gtf.gz \
| awk '$3 ~ /CDS/' \
| convert2bed --do-not-sort --input=gtf - \
> $exon_file
```

File: analyses/snv-callers/run_caller_consensus_analysis.sh [hunk 6]
```diff
@@ -45,15 +48,41 @@ Rscript analyses/snv-callers/scripts/02-merge_callers.R \
...
-  
+
+###################### Create intersection BED files ###########################
+# Make All mutations BED file
+bedtools intersect \
+  -a data/WGS.hg38.strelka2.unpadded.bed \
+  -b data/WGS.hg38.mutect2.unpadded.bed > $all_mut_wgs_bed
+
+# Convert GTF to BED file for use in bedtools
+# Here we are only extracting things labeled with CCDS ids, i.e. are coded in protein
```
Comment (jashapiro) [inline analyses/snv-callers/run_caller_consensus_analysis.sh]:
```suggestion
# Here we are only extracting lines tagged as CDS, i.e. are coded in protein
```

**Comment by cansavvy** [comment]
This has been cleared by @jashapiro after we've cleared up the `exon` issue in the GTF file. So it is ready for merging whenever.

**Pull request merged**